src/main/java/org/team1540/robot2025/subsystems/drive/Drivetrain.java

```diff
@@ -0,0 +1,424 @@
+                lastModulePositions = modulePositions;
+                lastOdometryUpdateTime = sampleTimestamps[i];
+            } else {
+                rejectedSamples++;
+            }
+        }
+        Logger.recordOutput("Odometry/RejectedSamples", rejectedSamples);
+
+        // Update robot velocities
+        ChassisSpeeds speeds = kinematics.toChassisSpeeds(getModuleStates());
+        speeds.omegaRadiansPerSecond =
+                gyroInputs.connected ? gyroInputs.yawVelocityRadPerSec : speeds.omegaRadiansPerSecond;
+        RobotState.getInstance().addVelocityData(speeds);
+
+        if (DriverStation.isEnabled()) {
+            // Run modules based on current drive mode
+            if (isFFCharacterizing) {
+                for (Module module : modules) module.runCharacterization(ffCharacterizationInput);
+            } else {
+                SwerveModuleState[] setpointStates;
+                if (OPTIMIZE_SETPOINTS) {
+                    SwerveSetpoint newSetpoint =
+                            setpointGenerator.generateSetpoint(lastSetpoint, desiredSpeeds, Constants.LOOP_PERIOD_SECS);
+                    setpointStates = newSetpoint.moduleStates();
+                    lastSetpoint = newSetpoint;
+                } else {
+                    setpointStates = kinematics.toSwerveModuleStates(
+                            ChassisSpeeds.discretize(desiredSpeeds, Constants.LOOP_PERIOD_SECS));
+                }
+
+                for (int i = 0; i < 4; i++) {
+                    modules[i].runSetpoint(setpointStates[i]);
+                }
+                Logger.recordOutput("Drivetrain/SwerveStates/Setpoints", setpointStates);
+            }
+        } else {
+            for (Module module : modules) module.stop(); // Stop modules when disabled
+            Logger.recordOutput(
+                    "Drivetrain/SwerveStates/Setpoints",
+                    new SwerveModuleState(),
+                    new SwerveModuleState(),
+                    new SwerveModuleState(),
+                    new SwerveModuleState());
+        }
+
+        // Update gyro alerts
+        gyroDisconnected.set(Robot.isReal() && !gyroInputs.connected);
+
+        // Update tunable numbers
+        LoggedTunableNumber.ifChanged(
+                hashCode(),
+                () -> trajectoryController.setTranslationPID(
+                        translationKP.get(), translationKI.get(), translationKD.get()),
+                translationKP,
+                translationKI,
+                translationKD);
+        LoggedTunableNumber.ifChanged(
+                hashCode(),
+                () -> {
+                    headingController.setPID(headingKP.get(), headingKI.get(), headingKD.get());
+                    trajectoryController.setHeadingPID(headingKP.get(), headingKI.get(), headingKD.get());
+                },
+                headingKP,
+                headingKI,
+                headingKD);
```

Comment:
Shouldn't this be only in tuning mode?

Reply:
`LoggedTunableNumber` does a check for tuning mode internally, on L71